src/primary_names.lua

```diff
@@ -0,0 +1,301 @@
+		name = name,
+		owner = owner,
+		startTimestamp = nameData.startTimestamp,
+		baseName = nameData.baseName,
+	}
+end
+
+---Finds all primary names with a given base  name
+--- @param baseName string -- the base name to find primary names for (e.g. "test" to find "undername_test")
+--- @return PrimaryNameWithOwner[]
+function primaryNames.getPrimaryNamesForBaseName(baseName)
+	local primaryNamesForArNSName = {}
+	local unsafePrimaryNames = PrimaryNames.names -- TODO: unsafe copy
+	for name, _ in pairs(unsafePrimaryNames) do
+		local nameData = primaryNames.getPrimaryNameDataWithOwnerFromName(name)
+		if nameData and nameData.baseName == baseName then
+			table.insert(primaryNamesForArNSName, nameData)
+		end
+	end
+	-- sort by name length
+	table.sort(primaryNamesForArNSName, function(a, b)
+		return #a.name < #b.name
+	end)
+	return primaryNamesForArNSName
+end
+
+---@class RemovedPrimaryName
+---@field owner WalletAddress
+---@field name ArNSName
+
+--- Remove all primary names with a given base  name
+--- @param baseName string
+--- @return RemovedPrimaryName[]
+function primaryNames.removePrimaryNamesForBaseName(baseName)
+	local removedNames = {}
+	local primaryNamesForBaseName = primaryNames.getPrimaryNamesForBaseName(baseName)
+	for _, nameData in pairs(primaryNamesForBaseName) do
+		local removedName = primaryNames.removePrimaryName(nameData.name, nameData.owner)
+		table.insert(removedNames, removedName)
+	end
+	return removedNames
+end
+
+--- Revoke claims created by a given process id
+--- @param initiator string -- the process id to revoke claims for, validated against the initiator of the claims
+--- @param names string[] -- the names to revoke claims for, if nil all claims for the initiator will be revoked
+--- @return PrimaryNameClaim[]
+function primaryNames.revokeClaimsForInitiator(initiator, names)
+	local revokedClaims = {}
+	names = names or utils.keys(PrimaryNames.claims)
+	for _, name in pairs(names) do
+		local claim = utils.deepCopy(PrimaryNames.claims[name])
+		if claim and claim.initiator == initiator then
+			PrimaryNames.claims[name] = nil
+			table.insert(revokedClaims, claim)
+		end
+	end
+	return revokedClaims
+end
+
+--- Get paginated primary names
+--- @param cursor string|nil
+--- @param limit number
+--- @param sortBy string
+--- @param sortOrder string
+--- @return PaginatedTable<PrimaryNameWithOwner>
+function primaryNames.getPaginatedPrimaryNames(cursor, limit, sortBy, sortOrder)
+	local primaryNamesArray = {}
+	local cursorField = "name"
+	for owner, primaryName in ipairs(PrimaryNames.owners) do
+		table.insert(primaryNamesArray, {
+			name = primaryName.name,
+			owner = owner,
+			startTimestamp = primaryName.startTimestamp,
+		})
+	end
+	return utils.paginateTableWithCursor(primaryNamesArray, cursor, cursorField, limit, sortBy, sortOrder)
+end
+
+--- Prune expired primary name claims
+--- @param timestamp number
+--- @return table<string, PrimaryNameClaim> the names of the claims that were pruned
```

Comment:
<img width="867" alt="image" src="https://github.com/user-attachments/assets/bff9e34b-2fc3-4703-90b0-c91587d158ee">
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/ebf366d2-5bbb-4d66-9407-21c19036da20">


this pattern will help IDE type support `---@return <type> [<name> [comment] | [name] #<comment>]`

```suggestion
--- Prune expired primary name claims
--- @param timestamp number
--- @return table<string, PrimaryNameClaim> prunedNameClaims the names of the claims that were pruned
```

Reply:
addressed a good chunk in [cebb3cb](https://github.com/ar-io/ar-io-network-process/pull/175/commits/cebb3cba577aa5a936939eb7bc99e01da74ebb9f)